src/libmongoc/src/mongoc/mongoc-ssl.c

```diff
@@ -178,5 +178,111 @@ _mongoc_ssl_opts_disable_ocsp_endpoint_check (const mongoc_ssl_opt_t *ssl_opt)
       ->tls_disable_ocsp_endpoint_check;
 }
 
+bool
+_mongoc_ssl_opts_from_bson (mongoc_ssl_opt_t *ssl_opt,
+                            const bson_t *bson,
+                            bson_string_t *errmsg)
+{
+   bson_iter_t iter;
+
+   if (ssl_opt->internal) {
+      bson_string_append (errmsg,
+                          "SSL options must not have internal state set");
+      return false;
+   }
+
+   ssl_opt->internal = bson_malloc0 (sizeof (_mongoc_internal_tls_opts_t));
+
+   if (!bson_iter_init (&iter, bson)) {
+      bson_string_append (errmsg, "error iterating BSON SSL options");
+      return false;
+   }
+
+   while (bson_iter_next (&iter)) {
```

Comment:
Not everyone likes the style I'm about to suggest, but I'll point out that if this is restructured a bit, some of the redundant error checking can vanish, and it also becomes clear that it could be separated into a couple of smaller functions IF that is seen as helpful. I believe the logic you have written here is fine, don't feel like you need to change this; but, submitted for your consideration is a suggestion. (Note: 0) This is /exactly/ the kind of thing that pattern matching REALLY helps with! 1) as can now be seen, splitting things into functions by "accepted data type" starts to look pretty good, but may just be overkill here.) 
```suggestion
   while (bson_iter_next (&iter)) {
      const char *key = bson_iter_key (&iter);

      if (BSON_ITER_HOLDS_UTF8 (&iter)) {
         if (0 == bson_strcasecmp (key, MONGOC_URI_TLSCERTIFICATEKEYFILE)) {
            ssl_opt->pem_file = bson_strdup (bson_iter_utf8 (&iter, NULL));
            return true;
         }

         if (0 ==
             bson_strcasecmp (key, MONGOC_URI_TLSCERTIFICATEKEYFILEPASSWORD)) {
            ssl_opt->pem_pwd = bson_strdup (bson_iter_utf8 (&iter, NULL));
            return true;
         }

         if (0 == bson_strcasecmp (key, MONGOC_URI_TLSCAFILE)) {
            ssl_opt->ca_file = bson_strdup (bson_iter_utf8 (&iter, NULL));
            return true;
         }

         // We got UTF8 data for something that doesn't take it:
         bson_string_append_printf (
            errmsg, "error, got unexpected UTF-8 value for key '%s'", key);
         return false;
      }

      if (BSON_ITER_HOLDS_BOOL (&iter))
      {
         if (0 ==
             bson_strcasecmp (key, MONGOC_URI_TLSALLOWINVALIDCERTIFICATES)) {
            /* If MONGOC_URI_TLSINSECURE was parsed, weak_cert_validation must
             * remain true. */
            ssl_opt->weak_cert_validation =
               ssl_opt->weak_cert_validation || bson_iter_bool (&iter);
            return true;
         }

         if (0 == bson_strcasecmp (key, MONGOC_URI_TLSALLOWINVALIDHOSTNAMES)) {
            /* If MONGOC_URI_TLSINSECURE was parsed, allow_invalid_hostname must
             * remain true. */
            ssl_opt->allow_invalid_hostname =
               ssl_opt->allow_invalid_hostname || bson_iter_bool (&iter);
            return true;
         }

         if (0 == bson_strcasecmp (key, MONGOC_URI_TLSINSECURE)) {
            if (bson_iter_bool (&iter)) {
               ssl_opt->weak_cert_validation = true;
               ssl_opt->allow_invalid_hostname = true;
            }

            return true;
         }

         if (0 == bson_strcasecmp (
                     key, MONGOC_URI_TLSDISABLECERTIFICATEREVOCATIONCHECK)) {
            ((_mongoc_internal_tls_opts_t *) ssl_opt->internal)
               ->tls_disable_certificate_revocation_check =
               bson_iter_bool (&iter);
            return true;
         }

         if (0 ==
             bson_strcasecmp (key, MONGOC_URI_TLSDISABLEOCSPENDPOINTCHECK)) {
            ((_mongoc_internal_tls_opts_t *) ssl_opt->internal)
               ->tls_disable_ocsp_endpoint_check = bson_iter_bool (&iter);
            return true;
         }

         // We got boolean data for something that doesn't take it:
         bson_string_append_printf (
            errmsg, "error, expected boolean value for key '%s'", key);
         return false;
      }
  
   bson_string_append_printf (errmsg, "unexpected option: %s", key);
   return false;
}
```

Reply:
Updated to check the value type first since that does make the error checks more concise. Replaced `return true` with `continue` since iteration must continue after successfully reading one field.

Reply:
Ah, DERP! Yes, continue is the right thing. :-)